---
title: "Week 1 - Hands-On Examples"
format: 
  html: default
  # pdf: default
categories: 
  - week01
  - exercise
editor_options: 
  chunk_output_type: console
---

The R script is available here:
[link](https://raw.githubusercontent.com/InforBio/IOC/refs/heads/main/ioc_r/week01/r_w01_exos.R)

## Create an R project

1. In RStudio, create a new project in your folder.

Some rules for naming your project:  

- be descriptive and keep it short
- use snake case (only lowercase letters and underscores allowed),
avoid special characters (such as `!`, `#`, ) and spaces
- the name cannot starts with numbers

2. Open your R project, create three new folders, *i.e.*, `data`, `scripts`, `outputs`.


## Import Dataset

### Data Description

We'll be working with a gene expression dataset as an example,
sourced from this [link](https://monashbioinformaticsplatform.github.io/r-more/).

The specific file we'll use is named "**read-counts.csv**",
which you can download from this
[zipped folder](https://monashbioinformaticsplatform.github.io/r-more/r-more-files.zip).


The data comes from an experiment using PCR to study 44 genes.
The results were measured to see which genes are active at different stages in Yeast cell cycling.
Several strains were tested, including wildtype and some with specific genes knock-downs.
Samples were taken at nine time points over two cell cycles (two hours).


### Importing into Rstudio

1. Download the zipped file to your computer and extract its contents.
2. If you are using the RStudio server, upload the `read-counts.csv` to the folder `data` of your R project.
3. Click on the file to "View" it and identify the column separator.
4. Import the file into R and call the imported data "counts".


## Play with Basic R Commands

Open an R script to write down your code. 
(In RStudio menu bar, click *File* -> *New File* -> *R Script*)

1. What is the absolute file path of the count data? What is its absolute path?
Verify your answer using the function `file.exists()`.

```{r}
#| echo: false
#| message: false

counts <- read.csv("../exos_data/read-counts.csv")
```

```{r}
#| eval: false

file.exists("relative_path/to/your/file") # replace the path by the yours
file.exists("absolute_path/to/your/file") # replace the path by the yours
```

2. Check the "Environment" panel or use the function `dim()`.
What is the dimension of the data frame?

```{r}
dim(counts)
```

In the "Environment" panel, click on the tabular icon next to the dataset to visualize the it.

We can extract all gene expressions for the sample named "WT:2" using `counts[["WT:2"]]`.

3. Try `mode()` on the expression data for "WT:2", what does it return?

```{r}
mode(counts[["WT:2"]])
```

4. Calculate the average expression (`mean()`) and standard deviation (`sd()`) of genes from the sample "WT:2".

```{r}
mean(counts[["WT:2"]])
sd(counts[["WT:2"]])
```

5. Generate descriptive statistics for all genes from the sample "WT:2" using `summary()`.

```{r}
summary(counts[["WT:2"]])
```

::: {.callout-tip title="Stats Time!"}
What are quartiles?

Quartiles are three values that split sorted data into four equal parts.

![](images/boxplot.png){width=65%}

(figure modified from this [source](https://www.google.com/url?sa=i&url=https%3A%2F%2Fcommons.wikimedia.org%2Fwiki%2FFile%3ABoxplot_vs_PDF.png&psig=AOvVaw3y4ZFGXe2JU0ypMGAEc2_i&ust=1734727102604000&source=images&cd=vfe&opi=89978449&ved=0CBQQjRxqFwoTCMDBxpfYtIoDFQAAAAAdAAAAABAE))

IQR (Interquartile range) = Q3 - Q1
:::


6. Generate a histogram for the "WT:2" sample using `hist()`. What does the distribution look like?

::: {.callout-tip title="Stats Time!"}
**Histograms** help us see how data is spread out.
They show how many data points fall into different ranges, or *bins*.
By looking at a histogram, we can quickly understand the shape of the data,
like if itâ€™s skewed or has outliers.
It's a simple way to get an overview of your data.
:::

```{r}
hist(counts[["WT:2"]])
```

::: {.callout-tip title="Stats Time!"}
**log2 Transformation**

A log2 transformation compresses larger values more than smaller ones,
making the data less skewed.
Many biological data, like gene expression counts,
tend to have a highly skewed distribution with a long tail of large values.
The log2 transformation makes the distribution more symmetric and closer to normal,
which is often desirable for statistical analysis
:::

```{r}
plot(
  x = counts[["WT:2"]], y = log2(counts[["WT:2"]]),
  xlab = "Original Scale", ylab = "log2 Scale",
  main = "Gene expression for the sample WT:2"
)
```

Re draw the histogram with the log2 transformed data,
what does the distribution look like now?

```{r}
hist(log2(counts[["WT:2"]]))
```


## Get Your First Quarto Report

Create a new Quarto document (*File* -> *New File* -> *Quarto Document*),
write all codes above and your observation for the questions in the document.
Then click the "Render" to generate your own report!

- Change something in your script Quarto and re-render it, is the report up-to-date?

To go further: 

- Where is your report stored?
- What should you do if you want the report be stored in a specified folder?
  => use a configuration file for Quarto.
  
Open a new text file and copy paste following code, save it as `_quarto.yml` in your project folder.

```r
project:
  output-dir: outputs/
```

Try "Render" again, now where is your report?


---

#### Good job! ğŸ‘ğŸ‘ You've taken your first big steps into R, and you're off to a great start, keep it up! 


