---
title: "Week 1 - Hands-On Examples"
category: "excercises"
---

## Some RStudio Settings

Go to *Tools* -> *Global Options...*

- Under the *General* tab:
  - Uncheck the box "Restore .RData..."
  - Set "Save workspace to .RData on exit" to **Never**.
  
![](images/rstudio_setting1.png)

- Under the *Code* tab:

In the *Display* panel, check the box "Use rainbow parentheses"

![](images/rstudio_setting2.png)
- Under the *Appearance* tab:

You can select a theme for you RStudio, e.g., the "Cobalt" for a dark theme.


- Under the *Pane Layout* tab:

You can arrange the four panels as you prefer.
In addition, click "Add Column" lets you open an extra "Source" code column, allowing you to view your scripts side by side!

![](images/rstudio_setting3.png)


## Create an R project

1. Create a new project in your folder.

Some rules for naming your project:  
- be descriptive and keep it short
- use snake case (only lowercase letters and underscores allowed), avoid special characters (such as `!`, `#`, ) and spaces
- the name cannot starts with numbers

2. Create three new folders: `data`, `scripts`, `outputs`


## Import Dataset

### Data

We will use a gene expression data as example.

The file is called "read-counts.csv" you can download it using this link: https://monashbioinformaticsplatform.github.io/r-more/r-more-files.zip

The data comes from an experiment using PCR to study 44 genes.
The results were measured to see which genes are active at different stages in Yeast cell cycling.
Several strains were tested, including wildtype and some with specific genes knock-downs.
Samples were taken at nine time points over two cell cycles (two hours).

Source: https://monashbioinformaticsplatform.github.io/r-more/topics/tidyverse.html


### Import into Rstudio

1. Download the zipped file and unzipped it
2. Upload the `read-counts.csv` to the folder `data` which you've just created
3. Click on the file to "View" it, what is the separator of each column?
4. Import the file into R and call it "counts"


## Play with basic R commands

1. What is the absolute file path of the count data?

```{r}
counts <- readr::read_csv("read-counts.csv")
```

2. What is the dimension of the data frame?

```{r}
dim(counts)
```

We can get the expression the the sample "WT:1" by using `counts[["WT:1"]]`.

3. Try `typeof()` and `mode()` on the expression data of "WT:1":

```{r}
typeof(counts[["WT:1"]])
mode(counts[["WT:1"]])
```

4. Calculate the average expression (`mean()`) and standard deviation (`sd()`) of genes from sample "WT:1".

```{r}
mean(counts[["WT:1"]])
sd(counts[["WT:1"]])
```

5. Generate a descriptive statistics of the data frame using `summary()`

```{r}
summary(counts[["WT:1"]])
```

6. Generate a histogram of the "WT:1" using `hist()` and what is the distribution?

```{r}
hist(counts[["WT:1"]])
hist(log2(counts[["WT:1"]]))
```

## Get Your First Quarto Report

Click the "Render" to generate the report!

- Change something in this script and re-render it, is the output up-to-date?
- Observe where is your report stored?
- What to do if you want the report be store in a specified folder?
  => use a configuration file for quarto.
  
Open a new text file and copy paste following code, save it as `_quarto.yml` in your project folder.

```
project:
  output-dir: outputs/
```

Try "Render" again, now where is your report?


---

Good job! You've taken your first big steps into R, and you're off to a great start, keep it up! üëèüëèüëè


